include/lbann/callbacks/callback_dump_outputs.hpp

```diff
@@ -0,0 +1,100 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2016, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef LBANN_CALLBACKS_CALLBACK_DUMP_OUTPUTS_HPP_INCLUDED
+#define LBANN_CALLBACKS_CALLBACK_DUMP_OUTPUTS_HPP_INCLUDED
+
+#include "lbann/callbacks/callback.hpp"
+
+namespace lbann {
+
+/** Dump layer output tensors to files.
+ *  This callback generates a text file for each output tensor of each
+ *  selected layer, computed at each mini-batch step. Each line
+ *  contains flattened tensor data corresponding to one mini-batch
+ *  sample (note that this is the transpose of the column-major matrix
+ *  representation we use internally). Output files are in the form
+ *  "<prefix><model>-<mode>-epoch<#>-step<#>-<layer>-output<#>.csv". This
+ *  is primarily intended as a debugging tool, although it can be used
+ *  for inference when performance is not critical.
+ *
+ *  Currently, only CSV output files are supported.
+ */
+class lbann_callback_dump_outputs : public lbann_callback {
+public:
+
+  /** Constructor.
+   *  @param layer_names    Names of layers with output dumps
+   *                        (default: dump outputs for all layers).
+   *  @param modes          Execution modes with output dumps
+   *                        (default: dump outputs for all modes).
+   *  @param batch_interval Frequency of output dumps (default: dump
```

Comment:
What is the choice for batch interval? I used the about to be deprecated dump activation a lot but mostly at every epoch end. Also, if this will support inference, one would expect the dump to be at forward_prop_end. I have no strong opinion but just want to know. One case I see against batch interval is that the user would have to compute number of iterations for an epoch if dumping at epoch boundary is desired (which mostly has been the way dump activations callback).

Reply:
The batch interval is used by the base callback class, so I don't think it's a good idea to hijack it to do something non-standard.

It's a good call to move the functionality to `on_forward_prop_end`/`on_evaluate_forward_prop_end`. I'll do that.